### ISSUE TITLE
Memory allocator crashes or spews infamous _set_color error

### ISSUE THREAD

**Issue by amedley** [godotengine/godot#37652]
**Godot version:** v3.2.1.stable.mono.official (64 bit)

**OS/device including version:** Windows 10 Home N (10.0.18362 Build 18362)

**Issue description:** The underlying custom allocator in Godot is having trouble with high-frequency allocations. I believe this is due to a race condition in `core/map.h`. Given a high frequency of allocs/deallocs (bound by a cap, see reproduction project) the allocator fails and the process either crashes or spews this infamous error:
> ERROR: _set_color: Condition "p_node == _data._nil && p_color == RED" is true.
> At: ./core/map.h:154

I say _infamous_ because there are few similar bug reports surrounding this evasive error, all of which involve a multi-threaded physics server. In the reproduction project linked below, you'll find my replication of this bug **without** any multi-threaded physics server.

Related & not duplicate:
https://github.com/godotengine/godot/issues/8630
https://github.com/godotengine/godot/issues/6512

**Steps to reproduce:**
1: (Recommended) Disable vsync to show greater discrepancy among idle_frame, _Process, and _PhysicsProcess. If vsync is enabled, the bug will take longer to reproduce.
2: Create a queue of thousands of object references, managed by Godot's custom allocator
3: Repeatedly dequeue old references and enqueue new references using any of Godot's life-cycle functions (fastest reproduction is in the idle frame, vsync disabled)
4: If you are using the idle frame (with vsync disabled), reproducing should only take a few seconds. View the logger to witness a crash or spewing of the infamous error. If you are using _Process or _PhysicsProcess, it may take a couple minutes, but the crash/spew always occurs.
5: (Optional): Build from source, set a breakpoint on map.h:154 and run the project. Assuming there is no crash, wait for the errors to spew. Once the spewing starts, attach the debugger and analyze the params. Look for condition `p_node == _data._nil && p_color == RED`, and check that it is truthy. If it is truthy, repeatedly step over, and you should see the "Finalizer" thread is stuck in an infinite loop as it attempts to balance the RB-Tree in `map.h`.

**Minimal reproduction project:** 
[reproduction-project.zip](https://github.com/godotengine/godot/files/4442831/reproduction-project.zip)

**Comment by amedley** [comment]
I got a crash dump backtrace which may help identify the underlying issue. As I suspected, something is amok in the allocator (specifically DefaultAllocator according to this backtrace)
**IMPORTANT: I modified map.h::_set_color** to help produce this particular crash dump backtrace. Here is the version I used specifically for producing the backtrace:
```
inline void _set_color(Element *p_node, int p_color) {
  if (p_node == _data._nil && p_color == RED) {
    // ERR_FAIL_COND(p_node == _data._nil && p_color == RED);
    _err_print_error(FUNCTION_STR, __FILE__, __LINE__, "p_node == _data._nil && p_color == RED");
  }
  p_node->color = p_color;
}
```
This just allows the color to be set even given the failed assertion. **This is not in the reproduction-project**, this is purely to help produce a rich backtrace. Here is what I got:

```
ERROR: Map<class Object *,struct CSharpScriptBinding,struct Comparator<class Object *>,class DefaultAllocator>::_set_color: p_node == _data._nil && p_color == RED
   At: D:\workspace\code\godot\core/map.h:155
ERROR: Map<class Object *,struct CSharpScriptBinding,struct Comparator<class Object *>,class DefaultAllocator>::_set_color: p_node == _data._nil && p_color == RED
   At: D:\workspace\code\godot\core/map.h:155
ERROR: Map<class Object *,struct CSharpScriptBinding,struct Comparator<class Object *>,class DefaultAllocator>::_erase_fix_rb: Condition "_data._nil->color != BLACK" is true.
   At: D:\workspace\code\godot\core/map.h:434
ERROR: Map<class Object *,struct CSharpScriptBinding,struct Comparator<class Object *>,class DefaultAllocator>::_set_color: p_node == _data._nil && p_color == RED
   At: D:\workspace\code\godot\core/map.h:155
ERROR: Map<class Object *,struct CSharpScriptBinding,struct Comparator<class Object *>,class DefaultAllocator>::_set_color: p_node == _data._nil && p_color == RED
   At: D:\workspace\code\godot\core/map.h:155
ERROR: Map<unsigned __int64,int,struct Comparator<unsigned __int64>,class DefaultAllocator>::_set_color: p_node == _data._nil && p_color == RED
   At: D:\workspace\code\godot\core/map.h:155
ERROR: Map<unsigned __int64,int,struct Comparator<unsigned __int64>,class DefaultAllocator>::_erase_fix_rb: Condition "_data._nil->color != BLACK" is true.
   At: D:\workspace\code\godot\core/map.h:434
ERROR: Map<unsigned __int64,int,struct Comparator<unsigned __int64>,class DefaultAllocator>::_erase: Condition "_data._nil->color == RED" is true.
   At: D:\workspace\code\godot\core/map.h:485
ERROR: Map<unsigned __int64,int,struct Comparator<unsigned __int64>,class DefaultAllocator>::_set_color: p_node == _data._nil && p_color == RED
   At: D:\workspace\code\godot\core/map.h:155
ERROR: Map<unsigned __int64,int,struct Comparator<unsigned __int64>,class DefaultAllocator>::_set_color: p_node == _data._nil && p_color == RED
   At: D:\workspace\code\godot\core/map.h:155
CrashHandlerException: Program crashed
Dumping the backtrace. Please include this when reporting the bug on https://github.com/godotengine/godot/issues
[0] Comparator<Object *>::operator() (D:\workspace\code\godot\core\typedefs.h:292)
[1] Comparator<Object *>::operator() (D:\workspace\code\godot\core\typedefs.h:292)
[2] Map<Object *,CSharpScriptBinding,Comparator<Object *>,DefaultAllocator>::_insert (D:\workspace\code\godot\core\map.h:339)
[3] Map<Object *,CSharpScriptBinding,Comparator<Object *>,DefaultAllocator>::insert (D:\workspace\code\godot\core\map.h:567)
[4] CSharpLanguage::insert_script_binding (D:\workspace\code\godot\modules\mono\csharp_script.cpp:1365)
[5] GDMonoInternals::tie_managed_to_unmanaged (D:\workspace\code\godot\modules\mono\mono_gd\gd_mono_internals.cpp:94)
[6] godot_icall_RandomNumberGenerator_Ctor (D:\workspace\code\godot\modules\mono\glue\mono_glue.gen.cpp:7947)
[7] <couldn't map PC to fn name>
[8] <couldn't map PC to fn name>
[9] <couldn't map PC to fn name>
[10] <couldn't map PC to fn name>
[11] <couldn't map PC to fn name>
```
Notice how the backtrace includes `godot_icall_RandomNumberGenerator_Ctor`, this is the mono glue for instancing a RandomNumberGenerator, which is the reference type I used in my reproduction project.

**Comment by akien-mga** [comment]
CC @neikeq

**Comment by MightyPrinny** [comment]
I have a similar error, but I can only reproduce it on my modern android phone, and for some reason I can't install the 64 bit app in that phone even though it has a cortex a53 processor. My old phone can install the the 32 and 64 bit app, but it doesn't have the error.

**Comment by trifox** [comment]
problem occurs when cleaning the stage, with lots of small items, 3.2.2 stable, unsure of exact pattern here

**Comment by MrZak-dev** [comment]
Same Issue occurs , 3.2.3 RC5 Mono 

![image](https://user-images.githubusercontent.com/27586088/92109534-dd10fa00-edf1-11ea-9bfe-6de23ef0434b.png)

**Comment by MrZak-dev** [comment]
@neikeq , @akien-mga

**Comment by neikeq** [comment]
Is this a C# only issue? Should I change the label?

**Comment by akien-mga** [comment]
@neikeq I don't know yet. The backtrace in https://github.com/godotengine/godot/issues/37652#issuecomment-610590537 does start in the Mono module, but that doesn't mean that this bug couldn't be triggered outside it (but it's a good place to start debugging I guess).

**Comment by MrZak-dev** [comment]
For more information .
This (may) start happening when the game start freeing a lot of instances at once .

**Comment by Avantir-Chaosfire** [comment]
Also using C#, also experiencing this problem. I had actually written my project in GDScript initially and did not encounter this issue at all. Shortly after converting to C# and making a small handful of changes, I'm running into this problem. It's certainly possible that I changed something that caused this, but it seems more likely to me that it's just not an issue with GDScript.

**Comment by Avantir-Chaosfire** [comment]
The reason I was encountering this error frequently was due to creating and matching a RegEx ~180 times per second, which is obviously unnecessary. This same code was running in GDScript for a long time and never had an issue, but it can hardly run a few minutes in mono without encountering this issue. I am very certain this issue is mono exclusive.

**Comment by RAREGUN** [comment]
The same problem! Sometimes when trying to create a large number of MeshInstance with PlaneMesh inside. I create them in 4 different threads, add them to the scene from the main one.
3.2.3.stable.mono

ERROR: _set_color: Condition "p_node == _data._nil && p_color == RED" is true.
   At: ./core/map.h:154

**Comment by ZackingIt** [comment]
Ditto -- I just hit this error just now as well. -- only clue I can offer to you is that it occurred after I changed a `call_deferred("add_child")` to a synchronous `add_child(....` I do not do any color manipulation in my game.

**Comment by omegachysis** [comment]
This looks like a race-condition or similar faulty synchronization at least on this part of the Godot C# subsystem:

https://github.com/godotengine/godot/blob/7e009a167110e79caa102c5b88363fd32b75293c/modules/mono/csharp_script.cpp#L1437-L1439

This gets called (without any locks, as far as I can tell) any time a mono-constructor is called on C# wrapper classes. The minimal project doesn't have explicit multithreading but I suspect that we are doing cleanup of dereferenced C# script objects and we don't protect against removals from the `script_bindings` member while we could be calling more constructors from a C# script.

I see this in my game all the time because I have a lot of allocations from C# scripts and I especially see it really bad when I have a background thread allocating C# wrapper objects (e.g. `new SurfaceTool()`) but then the game is closed, causing the cleanup routine to occur on all script objects while new ones are still being created.

Anyone think this is plausible? I get a lot of random CTDs in my game from lots of allocations like the one in the repro project.

**Comment by omegachysis** [comment]
I just tried adding a scoped mutex on `language_bind_mutex` around the offending line in `csharp_script.cpp`  and it appears that the issue no longer shows up in either the repro. project here or in my personal projects. Not sure if it'll have additional side effects but I'll submit a PR for it once I've tested it on `master`.

**Comment by omegachysis** [comment]
I can't reproduce this issue on `master`, so I think something was changed about the deallocation process in Godot 4.0 that fixes this. I'll open a 3.2-specific PR and maybe it could still be included in `master` because a mutex over it seems to be a good idea.

**Comment by MrZak-dev** [comment]
Are we getting any fixes for this in 3.2 stable version  ? 
I regret using C# for production because of this .

**Comment by MrZak-dev** [comment]
@neikeq

**Comment by peniwize** [comment]
I am using gdscript in my project (not C#) and some custom native C++ extensions and I'm encountering this same error.  After reading this post I suspect it's due to the [back end] C++ extensions instantiating several hundred Godot objects that are passed to the gdscript [front end], which it uses to allow the player to configure the game.  The Godot objects are all derived from Reference.  They're just simple data.  This error seems to have appeared after I added the new back end logic that generates several hundred Godot objects.

I'm also using an old version of Godot: git commit 978d71b8393ec425830ba48253dca4f32484edd1 from May 16 2019.  I'm aware of the reference counting bug in that version and have worked around it by delegating all object creation to gdscript helper methods so I don't think the C++ back end is doing anything wrong that way that could be causing a different error that ultimately causes the 'ERROR: _set_color: Condition "p_node == _data._nil && p_color == RED" is true.' error.

My primary reason for this post is to provide a data point proving that this issue is not exclusive to C#.  I believe the C# bindings and my custom bindings have similar behavior.  Perhaps other language bindings do/would cause this error as well.

**Comment by akien-mga** [comment]
Fixed by #43889.
